lib: Animate new items in ListingTable

marusak opened the pull request:
## Animate new rows in lists

When new items appear on a page, they are highlighted for a short while to spot them more easily.

![Peek 2022-03-15 07-47](https://user-images.githubusercontent.com/12330670/158322310-45f6838c-daf5-4a17-8181-6781a214635e.gif)

marusak commented on pkg/lib/cockpit-components-table.scss:
```diff
@@ -106,3 +106,28 @@
         border-bottom: 1px solid var(--ct-color-list-critical-border);
     }
 }
+
```
This might then go to some generic css file like page.css or so and then we can add the `.new-item` logic to other components and it would "just work" :tm:

marusak commented on pkg/lib/cockpit-components-table.jsx:
```diff
@@ -75,9 +75,36 @@ export const ListingTable = ({
+            return keys;
+        };
+
+        const new_keys = getRowKeys(rows);
```
This needs to be checked if it does not misbehave on any place where we use `ListingTable`. And it only works if the component specifies `props.keys` - but they should as it is a good practice if nothing else.
But I think that we might condition this by some property like `animateNewItems` so that we don't enable it by default everywhere.

marusak commented:
Without any other changes parts of storage page are already working:
![lvolumes](https://user-images.githubusercontent.com/12330670/157400133-09cb4ed2-c761-46d9-9c86-ea4996679b05.gif)

The side items ("Physical volumes", "Devices". "Drives"...) use `<SidePanel>` from `pkg/storaged/side-panel.jsx`. It should be possible to implement the same logic for it as well.

marusak commented:
> When visiting the SELinux page, the entries on screen are all marked as new.

Yeah, what I was worried about. As I mentioned above, we most likely want to enable this page by page and always make sure it works correctly. I briefly looked at selinux page and don't see any obvious issue (like using the same key), but I think it might be related how we fetch those alerts in the first place.

marusak commented:
Ah, the selinux is bit annoying - it reads all alerts without any details, renders then and then reads more info about them one by one and updates them. In this update a new key is generated. This whole seems a bit odd - both that we render it once and then update it item by item and that we generate it with different keys. If we decide to go with this PR I would just disable it on selinux page for now and try to fix this behaviour outside of this PR.

marusak commented:
@garrett I have fixed the issue on selinux page and also found one place that I also adjusted. In the first comment I dropped down list of all `ListingTable` occurrences and went through all of them.

I also moved the css parts into `pkg/lib/page.scss` so that we can re-use them in other components as well.

garrett commented on pkg/lib/page.scss:
```diff
@@ -76,29 +76,46 @@ a.disabled:hover {
-  }
-  70% {
-    background: #fdf4dd;
+}
```
So it starts with the background color, then doesn't do anything for half the time then fades out over 2 seconds.

The one below does a bunch of other stuff, but uses the same technique.

I adjusted the timing because it doesn't need to accommodate the moving animation time and then it also makes the animation slightly more gradual.

(Prefers reduced motion is not meant to completely remove all animations. It's meant to have little to no movement and not have drastic blinking. Mild, slow movements are even OK, although I went with just a simple fade for reduced motion.)

All major OSes have means to remove motion, usually under accessibility. I tested this on Firefox, which acknowledges GNOME's animation setting (in accessibility).

garrett replied:
Just explaining how this all works, especially on the default reduced motion, with more animation on the allows animation ('no preference") part.

martinpitt commented on pkg/lib/cockpit-components-table.jsx:
```diff
@@ -75,9 +75,36 @@ export const ListingTable = ({
+        if (currentRowsKeys.length !== 0) {
+            new_keys.forEach(key => {
+                if (currentRowsKeys.indexOf(key) === -1)
+                    setNewItems([...newItems, key]);
```
Is it wise to call a state set function that often? Shouldn't it rather do this once after the loop?

(And then use filtering on new_keys instead of forEach)

martinpitt commented on pkg/lib/cockpit-components-table.jsx:
```diff
@@ -148,6 +175,13 @@ export const ListingTable = ({
 
+        if (rowProps.key && newItems.indexOf(rowProps.key) >= 0) {
+            if (rowProps.className)
+                rowProps.className = rowProps.className + " new-item";
```
This is a globally visible class name, and thus should have a name which is less prone to collisions. But see below.

marusak replied:
Any suggestions? `new-item-animation` or you want to go one step further and add `cockpit` into mix as well e.g. `cockpit-item-animation`?

martinpitt replied:
`new-item-ct` seems good enough for now?

martinpitt commented on pkg/lib/cockpit-components-table.jsx:
```diff
@@ -148,6 +175,13 @@ export const ListingTable = ({
+            if (rowProps.className)
+                rowProps.className = rowProps.className + " new-item";
+            else
+                rowProps.className = "new-item";
```
Please drop the `if` and use `(rowProps.className || "")` for simplicity.

martinpitt commented on pkg/lib/page.scss:
```diff
@@ -75,6 +75,33 @@ a.disabled:hover {
 
+/* Animation of new items */
+
+.new-item {
```
Does this really have to go into the global page.css? I'd much rather like to see this in pkg/lib/cockpit-components-table.scss, to avoid namespace collisions. Otherwise please add a `ct-table` to the name. Same for `newRow`.

marusak replied:
I originally had it in `pkg/lib/cockpit-components-table.scss` but moved it to `pkg/lib/page.scss` as we want to use the same pattern on different components as well. As a good example see https://github.com/cockpit-project/cockpit/issues/10059 where we should also add animation but it does not use `ListingTable`. I purposefully want to create this as a global item so that we can just add this class into any page into any element and it just works the same way.

martinpitt replied:
I see, thanks! So let's please at least add a `-ct` suffix

martinpitt reviewed (changes requested):
Thanks, this is really cool! Animated transitions make it so much easier to see what's going on.

marusak commented:
@garrett thank you! looks good and I learned about the `prefers-reduced-motion` option as well

@martinpitt I have updated all your comments.

I also updated the gif in the first comment for the current changes. Not sure if we can have gif in release notes though.

martinpitt commented on pkg/lib/cockpit-components-table.jsx:
```diff
@@ -75,9 +75,37 @@ export const ListingTable = ({
+        if (currentRowsKeys.length !== 0) {
+            const new_keys = current_keys.filter(key => currentRowsKeys.indexOf(key) === -1);
+            if (new_keys.length)
+                setNewItems([...newItems, ...new_keys]);
```
This only ever gets added to, instead of just showing the actually new keys once rows changes. I know the animation will eventually fade out, but over time it would mean that all rows are "new"? Why not `setNewItems(new_keys)` here?

marusak replied:
Mostly I am worried that when two items would be added in quick succession that it would remove this new class from the freshly added item. I haven't tested what happens when you remove class mid animation but leaving the class there has no negative impact.

martinpitt reviewed (commented):
Thanks! One more question, but this is likely just my ignorance. (And will flip to +1 after answer)

martinpitt reviewed (approved):
Ack, thank you! Can you still prepare the release-notes bits, please?
